### ISSUE TITLE
[doc] Where do i find the finally trained model

### ISSUE THREAD

**Issue by stheid** [HumanCompatibleAI/imitation#249]
Is the `trainer.gen_algo` the finally trained model on which i can continue training?

I would like to clone a handcrafted controller into a stable-baselines agent and than continue training. It seems that the behavioural cloning algorithms are not related to stable-baselines at all, right?

I basically found your library though the sb3 documentation and thats also my main interest. https://stable-baselines3.readthedocs.io/en/master/guide/imitation.html

**Comment by shwang** [comment]
Hello, thanks for your interest in this repository.

Although this isn't officially supported by the API right now, `BC` (which is built on top of Stable Baselines 3 policies) can be used to fine-tune a SB3 policy with a few edits. Basically you'll need to modify these lines in `BC.__init__` to assign your preloaded instance of `sb3.common.policies.BasePolicy` to `self.policy` rather than X.

https://github.com/HumanCompatibleAI/imitation/blob/7c2882697b92b1b93cc32b36d4d4bce8916cf880/src/imitation/algorithms/bc.py#L201-L205

**Comment by shwang** [comment]
> Is the trainer.gen_algo the finally trained model on which i can continue training?

You should be able to fine-tune a SB3 algorithm with `GAIL` and `AIRL` (be wary that since porting to PyTorch, we've had a bug in the current version of AIRL, #229) by passing in any pretrained instance of `sb3.common.on_policy_algorithm.OnPolicyAlgorithm` (e.g. `PPO`, `TRPO`, etc).

I hope this answers your question.

**Comment by stheid** [comment]
thanks for your fast replay

> Although this isn't officially supported by the API right now, `BC` (which is built on top of Stable Baselines 3 policies) can be used to fine-tune a SB3 policy with a few edits. Basically you'll need to modify these lines in `BC.__init__` to assign your preloaded instance of `sb3.common.policies.BasePolicy` to `self.policy` rather than X.

actually i am interested in the opposite:
1. imitate a teacher agent
1. continue training with sb3 (hence i need a sb3 policy)

Hopefully i find time today to try out your suggestions.

My current understanding is, that IRL based imitation learning is not suitable for continued training as the Q-function, trained on the approximated reward function is off compared to the real reward function and the imitated agent will probably not learn properly on the real reward function in the beginning. But i am very new to imitation learning.

**Comment by shwang** [comment]
For GAIL/AIRL, as you suggested earlier, you can take `self.gen_algo` and use `self.gen_algo.set_env(ground_truth_reward_env); self.gen_algo.learn(n_timesteps)`. (By default, `self.gen_algo.env` is going to be a variant of the environment wrapped with GAIL or AIRL training reward).

`BC` will be a bit trickier, because it doesn't contain a SB3 `Algorithm` like PPO, but rather, contains a SB3 `Policy`. (SB3's serialization / deserialization API is built around Algorithms, not Policies). I don't know off the top of my head how to load a pretrained Policy into `PPO`, but I'm under the impression that you will need to do a bit of hacking and manual loading of weights.

**Comment by shwang** [comment]
> My current understanding is, that IRL based imitation learning is not suitable for continued training as the Q-function, trained on the approximated reward function is off compared to the real reward function and the imitated agent will probably not learn properly on the real reward function in the beginning. But i am very new to imitation learning.

I haven't tried this personally, but wouldn't be surprised if this were the case. When we've plotted the train reward function, it sometimes looks very different from the actual reward function.

You might be interested in [DQfD](https://arxiv.org/pdf/1704.03732.pdf). The 1st place winner of last year's MineRL competition added a trick where they adapted DQfD by linearly decreasing the number of demonstrations in a batch over time (in effect, smoothly transitioning from a "imitation learning" to pure Q learning).

Perhaps you could try an analogous thing where you linearly interpolate between the GAIL/AIRL reward and the ground truth reward.

**Comment by stheid** [comment]
```python
logger.configure(f"{timestamp} /BC/")
bc_trainer =BC(venv.observation_space, venv.action_space, expert_data=transitions, policy_class=ActorCriticPolicy)
bc_trainer.train(n_epochs=100)

class CopyPolicy(ActorCriticPolicy):
    def __new__(cls, *args, **kwargs):
        return bc_trainer.policy

model = sb3.PPO(CopyPolicy, venv, verbose=1)
model.save(f"{timestamp} /BC/checkpoint")
```
It's hacking :D I have not yet found time to validate that everything works, but the code does not crash so i assume it does some things correct.

Also your suggestions of gradually moving from Imitation learning to "normal" RL are cool, however, this again seems to invoke a lot of black magic, as you add another hyperparameter to tune (the speed of transition). Also playing a fraction of "real" actions can already be very instable. But the pointers are definetly welcome.

I would consider this issue closed for now. if i have results i might add them here, such that others can profit from my experiments. Thanks a lot
